include/crab/analysis/inter/top_down_inter_analyzer.hpp

```diff
@@ -931,6 +965,160 @@ analyze_function(CallGraphNode cg_node,
+      };
+
+  CRAB_LOG("inter-group", errs() << "Ins: ";
+           print_var_vector(errs(), src_params); errs() << "\n";);
+  CRAB_LOG("inter-group", errs() << "Outs: ";
+           print_var_vector(errs(), tgt_params); errs() << "\n";);
+
+  // Step 1. get all dsa intrinsic statements
+  const basic_block_t &entry_bb = tgt_cfg.get_node(tgt_cfg.entry());
+  std::vector<intrinsic_statement_t> intrinsic_stmts;
+  intrinsic_stmts.reserve(entry_bb.size());
+  for (auto const &stmt : entry_bb) {
+    if (stmt.is_intrinsic()) {
+      auto intrinsic_stmt = dynamic_cast<const intrinsic_statement_t &>(stmt);
+      if (intrinsic_stmt.get_intrinsic_name() == "regions_from_memory_object") {
+        intrinsic_stmts.push_back(intrinsic_stmt);
+      }
+    }
+  }
+
+  // Step 2. construct the equivalence classes for grouping input/output
+  // regions Note that, the order of the output regions in the parameter list
+  // is the same as the output list in the dsa intrinsic
+  // The worst case of the following loop is O(m * n) where m is the number of
+  // parameters and n is the number of dsa intrinsic statements
+  unsigned int index = 0, param_lst_sz = src_params.size();
+  while (index < param_lst_sz) {
+    if (!tgt_params[index].get_type().is_region()) {
+      // skip for reference and scalar variables
+      index++;
+      continue;
+    }
+    for (auto const &intrinsic_stmt : intrinsic_stmts) {
+      // each intrinsic contains only region variables
+      std::vector<Variable> srcs, tgts;
+      const unsigned intrinsic_arg_sz = intrinsic_stmt.get_num_args();
+      auto const &args = intrinsic_stmt.get_args();
+      srcs.reserve(intrinsic_arg_sz);
+      tgts.reserve(intrinsic_arg_sz);
+      if (intrinsic_arg_sz > 0 && index < param_lst_sz &&
+          args[0].get_variable() == tgt_params[index]) {
+        // If the region which the current index refers belongs to some dsa
+        // intrinsic, add this group into the equivalence sets
+        CRAB_LOG("inter-group", errs() << "Intrinsic: ";
+                 print_var_or_const_vector(errs(), args); errs() << "\n";);
+        for (unsigned j = 0; j < intrinsic_arg_sz; j++) {
+          srcs.push_back(src_params[index]);
+          tgts.push_back(tgt_params[index]);
+          index++;
+        }
+      }
+      if (srcs.size() > 0) {
+        src_cls.push_back(srcs);
+        tgt_cls.push_back(tgts);
+      }
+    }
+    index++;
+  }
+}
+
+// Wrapper to retrieve all dsa info from intrisinc statements.
+// The dsa info will be stored inside an object of class domains::callsite_info
+// for each callsite statement.
+// This analysis is performed before the inter-procedural analysis.
+// input: global_context
+// a. iterate all cfg inside wto of cfgs
+// b. for each cfg, visit callsite statement
+// c. for each callsite statement, get callee'cfg
+// d. store callsite info
+template <typename CallGraph, typename GlobalCtx>
+void analyze_callsite(const CallGraph &cg, GlobalCtx &ctx) {
```

Comment:
Don't call it `analyze_callsite`. This gives the impression that the callsite is actually analyzed by the abstract interpreter. Call it something like `extract_callsite_info` or similar